Issue #1273: CustomImportOrderCheck rule priorities

ivanov-alex opened the pull request:
Fixed. 
Also made few minor changes on Javadoc, mainly by removing `{@code }` for XML examples because it caused `&lt;` not being resolved to `<`

mkordas commented:
@ivanov-alex 
Not displayed violations on TeamCity are [known bug](https://youtrack.jetbrains.com/issue/TW-42425). In order to get green build just rebase over latest sources and remove unnecessary `this` from:
```java
            this.matchLength = length;
            this.matchPosition = position;
```

ivanov-alex commented:
@mkordas Thank you, it helped!
Do you know what could be the problem with AppVeyor?
```
[ERROR] Failed to execute goal org.codehaus.mojo:xml-maven-plugin:1.0:validate (default) on project checkstyle: 
While parsing C:\projects\checkstyle\src\main\resources\sun_checks.xml: null: UndeclaredThrowableException: Connection reset -> [Help 1]
```
This is definitely not a something I have changed...

mkordas commented:
`Connection reset`, so probably something with AppVeyor itself. I've just retriggered the build and we'll see.

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java:
```diff
@@ -524,4 +522,58 @@ public void testDefaultConfiguration() throws Exception {
+        final String[] expected = {
+            "9: " + getCheckMessage(MSG_ORDER, THIRD, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.JavadocNodeImpl"),
+            "21: " + getCheckMessage(MSG_NONGROUP_EXPECTED, STD, "com.puppycrawl.tools.checkstyle.checks.javadoc.WriteTagCheck"),
+            //"24: " + getCheckMessage(MSG_NONGROUP_EXPECTED, SPECIAL, "com.puppycrawl.tools.checkstyle.checks.javadoc.HtmlTag"),
```
please remove commented out code

mkordas commented on src/xdocs/config_imports.xml:
```diff
@@ -825,7 +825,55 @@ import java.util.regex.Matcher; //#8
+        To set Regexps for THIRD_PARTY_PACKAGE and STANDARD_JAVA_PACKAGE groups use
+        thirdPartyPackageRegExp and standardPackageRegExp options.
+        </p>
+ <p>
```
please fix indentation from this point on

mkordas commented on src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/imports/InputCustomImportOrder_MultiplePatternMatches.java:
```diff
@@ -0,0 +1,8 @@
+package com.puppycrawl.tools.checkstyle.imports;
```
is there any reason for it to be in non-compilable resources?

ivanov-alex replied:
Will it compile with non-existing "import org.myOrgorgan.Test;"?

mkordas replied:
@ivanov-alex sorry, maybe I'm missing something, but why do we need non-existing import for that test? Can't we create file that uses only existing imports?

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:
```diff
@@ -98,6 +100,51 @@
  * thirdPartyPackageRegExp and standardPackageRegExp options.
  * </p>
+ * <p>
+ * Pretty often one import can match more then one group. For example, static import from standard
```
`then` -> `than`

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:
```diff
@@ -98,6 +100,51 @@
+ * In this case, group will be assigned according to priorities:
+ * </p>
+ * <p>
+ *    1) STATIC has top priority
```
I think here we should use ordered list (`<ol>`)

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.java:
```diff
@@ -98,6 +100,51 @@
+ *    4) THIRD_PARTY has the least priority
+ * </p>
+ * <p>
+ *Few examples to illustrate "best match":
```
I believe that we should keep space between `*` and contents in Javadoc

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java:
```diff
@@ -524,4 +522,58 @@ public void testDefaultConfiguration() throws Exception {
+    }
+
+    @Test
+    public void testMultiplePatternMatches2() throws Exception {
```
please do not use numbered tests - each test should have its own name telling the exact thing it's going to check

ivanov-alex replied:
This test covers one of the options of "overlapping patterns": patterns of the same length, where match for second pattern on a configuration goes first on input file. Adding meaningful abbreviation of that to method name is not that simple.

mkordas replied:
let's try, I don't mind even if method name will be 80 characters

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheckTest.java:
```diff
@@ -524,4 +522,58 @@ public void testDefaultConfiguration() throws Exception {
+            createCheckConfig(CustomImportOrderCheck.class);
+        checkConfig.addAttribute("customImportOrderRules",
+                "SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE");
+        checkConfig.addAttribute("specialImportsRegExp", "myO");
```
can we use some more meaningful pattern here than `myO`?

ivanov-alex replied:
@mkordas changed to ".my". Is it good enough or you mean some scenario that could be closer to real life (it will be more difficult to find one comparing to synthetic test above)?

mkordas commented on src/xdocs/config_imports.xml:
```diff
@@ -825,7 +825,55 @@ import java.util.regex.Matcher; //#8
+        </p>
+ <p>
+ Pretty often one import can match more then one group. For example, static import from standard
+ package or regular expressions are configured to allow one import match multiple groups).
```
`groups)` - why do we need closing parenthesis here?

ivanov-alex commented:
@mkordas Fixed everything except of 2 comments where I have provided my feedback inline.

mkordas commented:
@ivanov-alex, comments answered inline

ivanov-alex commented:
@mkordas Done

romani commented:
@ivanov-alex , please prove by reports that there are no difference in reports for Guava before and after fix, or if there is a diff - explain it.

ivanov-alex commented:
I have added few more google projects to report because Guawa is clean.
Reports ([before](http://ivanov-alex.github.io/i1273/target_google_before/site/checkstyle.html), [after](http://ivanov-alex.github.io/i1273/target_google_after/site/checkstyle.html)) [are identical](https://www.diffchecker.com/hw6pesau)

I have also rebased the code.

mkordas commented:
@ivanov-alex, thanks for the reports.

1. What was the reason to remove space just after star in Javadoc? It renders fine with this space.
2. Why have you removed `@code`? They are really useful for storing XML. You should rather replace `&lt;` with `<` for better readability.
3. Sometimes you use `Regexp`, sometimes `RegExp` form. Please be at least consistent.

ivanov-alex commented:
>    What was the reason to remove space just after star in Javadoc? It renders fine with this space.

I have removed space in `@code` blocks because it renders as is (i.e. with space on the line beginning)
For example
```
*{@code
*import com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck;
*import com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderCheck;}
```
renders as below (try selecting the piece of code to see the space)
```
 import com.puppycrawl.tools.checkstyle.checks.imports.CustomImportOrderCheck;
 import com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderCheck;
```

>    Why have you removed @code? They are really useful for storing XML. You should rather replace &lt; with < for better readability.

Personally, I dislike these `@code` annotations in this project. It doesn't have any preferences (like markup), it just renders something as is.
We are supporting two sources javadoc and xdoc. I would prefer to have them as similar as possible. I still should prepare HTML-friednly version for xdoc, so why should I reformat everything for javadoc, which also works fine with html. 
My personal preference is to have them identical (with the only difference in space-star-space on every line for javadoc). It will reduce efforts and will help us have javadoc and xdoc consistent - the more deviations between them, the harder to have them consistent.
>    Sometimes you use Regexp, sometimes RegExp form. Please be at least consistent.

Done

romani commented:
> My personal preference is to have them identical

good point ! Eventually we will did converter from javadoc to xdoc, to let us manage this documentation only in one place - javadoc

romani commented:
merged as FF
